volatilityfoundation/volatility3
# [2024] AArch64 support

Hi :wave:,

This PR provides AArch64 integration to the Volatility3 framework, as well as a design rework in the current linux stacker. 

Implementation follows ARM official documentation, and includes essential APIs for higher level code (plugins etc.).

You can follow the roadmap here : 
- https://github.com/volatilityfoundation/volatility3/issues/161
---

My ressources : 
- https://gist.github.com/Abyss-W4tcher/f1833623c975193446315d48c106750e
- https://gist.github.com/Abyss-W4tcher/8442b6b6b85f725158fe7e9b99e507be
---

Testing : 

- RaspberryPi (virtual) : `Linux version 6.1.21`
- Android AVD (virtual) : `Linux version 3.18.94`
- RaspberryPi (physical) : `Linux version 6.1.0-rpi8-rpi-v8`
- Google Pixel 6A (physical) : `Linux version 5.10.157-android13-4-00003-g830b023b88f3-dirty`
- Jetson lime (physical) : `Linux version 4.9.201-tegra`
- NXP LX2160 (physical) : `Linux version 6.1.36`

_Thanks to everyone who took the time to test this PR on their devices !_

---

Unstable plugins : 

- `linux.bash` : 
     - `sections have no size` on Android related samples. 
     - 12/04/24 update : Android does not ship with `bash` by default, as [it's not GNU/Linux running on the system ](https://www.reddit.com/r/AskTechnology/comments/z2vqyt/comment/ixkaf1l/). Either we need a dedicated plugin (`.ash_history` ?), or a check to tell users why the plugin isn't working.
- `linux.lsof` : 
     - `dentry` members can be NULL, and this behaviour isn't handled (more informations in the dedicated Slack thread)

---

Please note that this is still experimental, testing is still going on. If you want to try this PR out, here are the steps : 

- Merge/checkout the PR into your local Volatility3 copy (`git checkout aarch64-support`)
- Make a memory capture of your AArch64 device (see attached ressources)
- Create the ISF against the vmlinux file with `dwarf2json`
- Set the following on top of the plugins you plan to use : 
```python
requirements.ModuleRequirement(
      name="kernel",
      description="Linux kernel",
      architectures=["Intel32", "Intel64", "AArch64"],
)
```

If you have any *real-life* samples to provide, or if you encounter any error, feel free to comment below !
